controllers/purge_controller.go

```diff
@@ -0,0 +1,209 @@
+type PurgeReconciler struct {
+	client.Client
+	record.EventRecorder
+	RequeueIntervals
+	signature.VerificationSettings
+	SKRWebhookManager     watcher.SKRWebhookManager
+	KcpRestConfig         *rest.Config
+	RemoteClientCache     *remote.ClientCache
+	ResolveRemoteClient   RemoteClientResolver
+	PurgeFinalizerTimeout time.Duration
+}
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *PurgeReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := ctrlLog.FromContext(ctx)
+	logger.V(log.InfoLevel).Info("Purge Reconciliation started")
+
+	ctx = adapter.ContextWithRecorder(ctx, r.EventRecorder)
+
+	// check if kyma resource exists
+	kyma := &v1beta1.Kyma{}
+	if err := r.Get(ctx, req.NamespacedName, kyma); err != nil {
+		if apierrors.IsNotFound(err) {
+			remoteClient, err := r.ResolveRemoteClient(ctx, client.ObjectKey{Namespace: req.Namespace, Name: req.Name})
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+			if done, msg := performCleanup(ctx, remoteClient, logger); done {
+				logger.Info(fmt.Sprintf("%s", msg))
+			} else {
+				logger.Info("Purging was not successful!")
+				logger.Info(fmt.Sprintf("%s", msg))
+				//TODO: Should we requeue? If the error is permanent, we may end up in the endless loop...
+			}
+
+			return ctrl.Result{}, nil
+		}
+
+		return ctrl.Result{}, err
+	}
+
+	// condition to check if deletionTimestamp is set, retry until it gets fully deleted
+	if !kyma.DeletionTimestamp.IsZero() || kyma.Status.State == v1beta1.StateDeleting {
+		deletionDeadline := kyma.DeletionTimestamp.Add(r.PurgeFinalizerTimeout)
+
+		if time.Now().After(deletionDeadline) {
+			remoteClient, err := r.ResolveRemoteClient(ctx, client.ObjectKeyFromObject(kyma))
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+			if done, msg := performCleanup(ctx, remoteClient, logger); done {
+				logger.Info(fmt.Sprintf("%s", msg))
+			} else {
+				logger.Info("Purging was not successful!")
+				logger.Info(fmt.Sprintf("%s", msg))
+			}
+			return ctrl.Result{}, nil
+		}
+
+		return ctrl.Result{
+			//	Requeueing exactly after 1s of the deadline
+			RequeueAfter: deletionDeadline.Add(time.Second).Sub(time.Now()),
+		}, nil
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// Helper functions
+func performCleanup(ctx context.Context, remoteClient client.Client, logger logr.Logger) (done bool, msg interface{}) {
+	var crdList = apiextensions.CustomResourceDefinitionList{}
+
+	if err := remoteClient.List(ctx, &crdList); err != nil {
+		return false, err
+	}
+
+	for _, crdResource := range crdList.Items {
+		staleResources, err := getStaleResourcesFrom(ctx, remoteClient, crdResource)
+		//	Skipping Kyma Custom Resource deletion as it should be handled by the main reconciler loop
+		if crdResource.Spec.Group == "operator.kyma-project.io" && crdResource.Spec.Names.Kind == "Kyma" {
```

Comment:
You wrap this condition check in a func `isKymaCR` and even do the skip before getting the stale resources